design/one-pager-moving-packages.md

```diff
@@ -0,0 +1,462 @@
+    - provider: crossplane-contrib/provider-dependency-a
+      version: ">= v1.0"
+    - provider: provider-dependency-b
+      version: ">= v1.0"
+```
+
+When you install that package, the revision controller adds the following entry
+to the Lock:
+
+```yaml
+apiVersion: pkg.crossplane.io/v1beta1
+kind: Lock
+metadata:
+  name: lock
+packages:
+- name: crossplane-contrib-configuration-example-26746e8e7abb
+  type: Configuration
+  source: xpkg.upbound.io/crossplane-contrib/configuration-example
+  version: v0.2.0
+  dependencies:
+  - constraints: ">= v1.0"
+    package: xpkg.upbound.io/crossplane-contrib/provider-dependency-a
+    type: Provider
+  - constraints: ">= v1.0"
+    package: xpkg.upbound.io/crossplane-contrib/provider-dependency-b
+    type: Provider
+```
+
+The registry for provider-dependency-a is copied from the dependent package.
+Similarly, the registry and organization for provider-dependency-b is copied
+from the dependent.
+
+Now assume `xpkg.upbound.io/crossplane-contrib/configuration-example` you
+replicate to `pkg.acme.co/internal/provider-dependency-a` - an internal
+registry.
+
+When you install that package, the revision controller adds the following entry
+to the Lock:
+
+```yaml
+apiVersion: pkg.crossplane.io/v1beta1
+kind: Lock
+metadata:
+  name: lock
+packages:
+- name: internal-configuration-example-26746e8e7abb
+  type: Configuration
+  source: pkg.acme.co/internal/configuration-example
+  version: v0.2.0
+  dependencies:
+  - constraints: ">= v1.0"
+    package: pkg.acme.co/crossplane-contrib/provider-dependency-a
+    type: Provider
+  - constraints: ">= v1.0"
+    package: pkg.acme.co/internal/provider-dependency-b
+    type: Provider
+```
+
+Again, the registry for provider-dependency-a is copied from the dependent
+package. The registry and organization for provider-dependency-b is
+copied from the dependent.
+
+Dependencies can still be fully qualified. This may be desirable - for example
+for internal packages. Fully qualifying a dependency ensures it'll always be
+installed from your internal registry.
+
+Crossplane uses the [go-containerregistry] to parse OCI references. It
+configures the library to use `xpkg.crossplane.io` as the default registry. You
+can override this by passing the `--registry` flag to Crossplane.
+
+At the time of writing, Crossplane does support partially qualified
+dependencies, but not in the proposed way. Currently:
+
+* If the registry is omitted, it's defaulted to `xpkg.upbound.io`
+* If the organization is omitted, it's defaulted to `library`
+
+This means defaulting to the dependent's registry and organization would be a
+breaking behavior change. A partially qualified dependency would have a
+different meaning before and after the change.
+
+We have two tools available to manage this breaking change:
+
+1. Put it behind a feature flag - e.g. `--enable-derive-dependencies-from-parent`
+2. Use a package's `spec.crossplane.version` to restrict it to new versions
```

Comment:
I realized this late in the game. It's a bit concerning.

It's important to support is letting folks replicate the big provider families (e.g. `upbound/provider-family-aws` and its dependents) into their own private registries, for example if their security or resiliency policy requires it.

One approach we could take is to:

* Put the behavior change behind a feature flag
* Don't make the provider's dependencies on provider-family-aws partially qualified until the flag is on by default
* Use `spec.crossplane.version` to restrict them to the Crossplane version where the flag is on by default

However I imagine this wouldn't be great for the community. They'd need to update to the latest Crossplane, or be stuck on old provider versions.

Perhaps we could take a long term view here. Make the change, in Crossplane, but not update any major existing providers to use partially qualified dependencies until the behavior was enabled by default in every supported Crossplane version?

Reply:
The other thing to consider is forward compatibility. Any packages out there today with partially qualified dependencies would need to be updated before the flag was enabled by default.

Reply:
Wondering if we could have a 3rd option by getting this from package metadata 🤔 
Similar to the replaces stanza, imagine we have a dedicated field like `spec.deriveDependenciesFromParent` defaulting to false.

For old packages + new crossplane: the behavior stays same as today.
For new packages + old crossplane: the behavior stays same as today.
For new packages + new crossplane: we get the new behavior.

Reply:
@turkenh That crossed my mind too. I don't think it would work for new packages + old Crossplane though. Or am I missing something?

I'm assuming:

* `spec.deriveDependenciesFromParent` is true
* `spec.dependencies` are partially qualified - e.g. `crossplane-contrib/provider-family-aws`

The old version of Crossplane won't be aware of the `spec.deriveDependencesFromParent` field in the package metadata, right? So it'll always treat partially qualified dependencies as meaning:

* If the registry was omitted, use the default registry set by `--registry`
* If the organization was omitted, use `library`

So if you've set `--registry` to `internal.acme.co` it'll try pull from there. That could be quite disruptive if it was e.g. provider-family-aws that was successfully pulling from a fully qualified `xpkg.upbound.io/crossplane-contrib/provider-family-aws`.

Reply:
Oh - are you thinking `spec.dependencies` would remain fully qualified, and `spec.deriveDependenciesFromParent` would cause segments (e.g the registry and/or org) to be replaced?

Intuitively that feels more to me like an install-time concern, since it's now overriding the specified dependencies (as opposed to making them less specific). So I'd probably expose it as a knob on the `pkg.crossplane.io` object rather than the package metadata (`meta.pkg.crossplane.io`).

It feels like a variant of the "Rewrite Dependency Prefixes" that @phisco proposed, only without the prefix matching. Something like this:

```yaml
apiVersion: pkg.crossplane.io/v1
kind: Provider
metadata:
  name: provider-aws-s3
spec:
  package: internal.acme.co/mirror-upbound/provider-aws-s3:v1.5.0
  rewriteDependencies: RegistryFromDependent
```

The options for `rewriteDependencies` would be something like:

* `RegistryFromDependent` - e.g. replace `xpkg.upbound.io` -> `internal.acme.co`
* `RegistryAndOrganizationFromDependent` - e.g replace `xpkg.upbound.io/upbound` -> `internal.acme.co/mirror-upbound`
* `DoNotRewrite` (default) - Don't rewrite dependencies

Reply:
I'll wait to hear what others think, but I'm leaning toward updating this design to recommend the `rewriteDependencies` approach I mentioned in my last comment.

If this were a greenfield design I'd still lean toward the partially qualified dependencies approach. It's not though, and `rewriteDependencies` has a bunch of nice properties:

* It's compatible with all existing packages - no need to wait for a bunch of package metadata updates
* It's pretty explicit/unsurprising - it's right there on the package spec
* It doesn't rely on prefix matching, so it shouldn't break if a package changes its dependencies

One challenge though is that we'd presumably need to propagate the setting down - i.e. ensure when we create dependencies they have the same `rewriteDependencies` setting.

Reply:
> Oh - are you thinking spec.dependencies would remain fully qualified, and spec.deriveDependenciesFromParent would cause segments (e.g the registry and/or org) to be replaced?

Not really, but the former, but your explanation made it clear that it has consequences. 

> I'm leaning toward updating this design to recommend the rewriteDependencies approach I mentioned in my last comment.

What happens if two packages have a common dependency with different `rewriteDependencies` setting / or somehow results in different rewritten image sources?

Reply:
> What happens if two packages have a common dependency with different rewriteDependencies setting / or somehow results in different rewritten image sources?

There's potential for a conflict, but I'm not sure it's much worse than the potential that already exists today.

I can think of a couple of scenarios:

1. `xpkg.upbound.io/pkg/a` is already installed. Someone tries to install a package that depends on it, but rewrites its registry to `acme.co/pkg/a`.
1. `xpkg.upbound.io/pkg/a` is already installed. Someone tries to install a package that depends on it, but rewrites its registry and organization to `acme.co/packages/a`.

There's also the reverse of these scenarios - i.e. the rewritten version is installed first, then someone tries to install the original / authoritative version second.

In scenario 1:

1. The new source (`acme.co/pkg/a`) will be added to the Lock as a dependency
2. The resolver controller will try to create a package (e.g. provider) named `pkg-a`, which will silently fail because it already exists
3. The new `source` will thus never be added as a top-level package
4. The package that depends on `acme.co/pkg/a` will thus never have its dependency satisfied and never finish installing

In scenario 2:

1. The new source (`acme.co/pkg/a`) will be added to the Lock as a dependency
2. The resolver controller will try to create a package (e.g. provider) named `packages-a`, which will work because no such package exists
3. The new `source` will be added as a top-level package, thus allowing the package that depends on `acme.co/packages/a` to finish installing
4. The revision controller will try to install the package's payload, which will fail due to the controller revision conflict, so the new revision and provider will never become healthy

Neither of these are great, but similar issues already exist today if:

* You install a package from xpkg.upbound.io and then mirror it and manually install the mirrored package
* You install two different providers that happen to share the same org and repo names (in different registries)
* You install two different providers that want to own the same CRDs

I'd argue that if you're going to mirror Crossplane packages, it requires some discipline as a control plane administrator to make sure you always specify a `rewriteDependencies` setting to ensure dependencies are pulled from where you expect them to be pulled.

Reply:
> However, if the --registry flag is set to index.docker.io (for example), it might break. Before the change the dependency resolves to xpkg.upbound.io/upbound/provider-family-aws. After the change it resolves to index.docker.io/upbound/provider-family-aws. That package probably won't exist, so pulls will stop working.

Maybe we can live with this?

Assume we:

* Change Crossplane's behavior so that partially qualified dependencies are relative to their parent
* Update provider families to use partially qualified dependencies (e.g. `upbound/provider-family-aws`)

For anyone running with the default `--registry` flag and installing packages from `xpkg.upbound.io`, the partially qualified dependencu will resolve to `xpkg.upbound.io/upbound/provider-family-aws` whether you're running an old version of Crossplane or a new one.

The only people affected would be folks running old versions of Crossplane:

1. Who are running with a non-default `--registry` flag or...
2. Who try to consume a package that's updated to use a partially qualified dependency **without** an org (since Crossplane would try pull from `library`)

We could reduce the impact of 2 by strongly advising folks not to drop the org from their dependencies - at least until Crossplane v1.21 is released (when v1.17 would no longer be supported).

That leaves only folks who are:

* Running an older version of Crossplane
* Using a non-default `--registry` flag
* Trying to install packages from xpkg.upbound.io that have switched to partially qualified dependencies

Maybe this set of folks is small enough that we could give notice of the breaking change without a major disruption?